### PR TITLE
Fix ValueError: I/O operation on closed pipe error on Windows 10

### PR DESCRIPTION
**Pull request by AmonUCF**
This fix references [this issue](https://github.com/Lux-AI-Challenge/Lux-Design-2022/issues/80). I don't fully understand what's going on to cause the issue, but it seems that without properly waiting for subprocesses to terminate, we get errors from pending I/O operations. I went ahead and added a cleanup method to be run at the end of execution for an episode, which seems to resolve the issue on windows. 

With better understanding of the bug, someone might be able to make a cleaner solution this what I have here.

### CODE REVIEW

**Comment by StoneT2000** [comment]
Thanks for contributing! I’ll see if a few others can test this out. Code wise it looks fine.

**Comment by AmonUCF** [comment]
So I had one "issue" with this fix which I noticed today. After terminating the subprocesses, we'd get empty error logs being emmited somehow. This would just end up having the console look like this after completion:

```
...
98: Invalid Dig Action for unit [0] unit_6 UnitType.HEAVY at (15, 41) - Tried to dig on top of a factory
99: Invalid Dig Action for unit [0] unit_6 UnitType.HEAVY at (15, 41) - Tried to dig on top of a factory
100: Invalid Dig Action for unit [0] unit_6 UnitType.HEAVY at (15, 41) - Tried to dig on top of a factory
120: player_1 lost all factories
player_0, .\kits\python\main.py: 
player_1, .\kits\python\main.py: 
0.4700615406036377
```

As you can see, both player 0 and 1 emit empty error logs. I've added a check in the process.py to see if we're at the end of the stream and only emit if the error log is non-zero in length.

File: luxai_runner/process.py [hunk 1]
```diff
@@ -91,4 +92,8 @@ async def stderr(self):
         #         # standard error output received, print it out
         #         stderrs.append((await line).decode())
         #         # stderrs.append(line.decode())
-        # return " ".join(stderrs)
\ No newline at end of file
+        # return " ".join(stderrs)
+
+    async def cleanup(self):
+        self._agent_process.terminate()
+        await self._agent_process.wait()
```
Comment (jmerle) [inline luxai_runner/process.py:99]: On Ubuntu 22.04 with Python 3.8.13 I'm seeing these errors after running a match with the changes in this PR:
```
$ luxai2022 build/agents/v01/main.py build/agents/v01/main.py
200: player_0 lost all factories
200: player_1 lost all factories
1.3537261486053467
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x7fce2e90f5e0>
Traceback (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 126, in __del__
    self.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 104, in close
    proto.pipe.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 536, in close
    self._close(None)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 560, in _close
    self._loop.call_soon(self._call_connection_lost, exc)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x7fce2e90f5e0>
Traceback (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 126, in __del__
    self.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 104, in close
    proto.pipe.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 536, in close
    self._close(None)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 560, in _close
    self._loop.call_soon(self._call_connection_lost, exc)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
```

I haven't found a fix yet, but the error stops appearing when I comment out this line.

Comment (jmerle) [reply luxai_runner/process.py:99]: For some reason the stdin/stdout/stderr pipes are getting closed after the event loop closes. Full debug logs:
```
$ python -W default -X tracemalloc -m luxai_runner.cli build/agents/v01/main.py build/agents/v01/main.py
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:asyncio:execute program 'python' stdin=<pipe> stdout=<pipe> stderr=<pipe>
DEBUG:asyncio:process 'python' created: pid 92428
DEBUG:asyncio:execute program 'python' stdin=<pipe> stdout=<pipe> stderr=<pipe>
DEBUG:asyncio:process 'python' created: pid 92430
DEBUG:asyncio:Write pipe 8 connected: (<_UnixWritePipeTransport fd=8 idle bufsize=0>, <WriteSubprocessPipeProto fd=0 pipe=<_UnixWritePipeTransport fd=8 idle bufsize=0>>)
DEBUG:asyncio:Write pipe 10 connected: (<_UnixWritePipeTransport fd=10 idle bufsize=0>, <WriteSubprocessPipeProto fd=0 pipe=<_UnixWritePipeTransport fd=10 idle bufsize=0>>)
DEBUG:asyncio:Read pipe 9 connected: (<_UnixReadPipeTransport fd=9 polling>, <ReadSubprocessPipeProto fd=1 pipe=<_UnixReadPipeTransport fd=9 polling>>)
DEBUG:asyncio:Read pipe 12 connected: (<_UnixReadPipeTransport fd=12 polling>, <ReadSubprocessPipeProto fd=1 pipe=<_UnixReadPipeTransport fd=12 polling>>)
DEBUG:asyncio:Read pipe 11 connected: (<_UnixReadPipeTransport fd=11 polling>, <ReadSubprocessPipeProto fd=2 pipe=<_UnixReadPipeTransport fd=11 polling>>)
DEBUG:asyncio:Read pipe 14 connected: (<_UnixReadPipeTransport fd=14 polling>, <ReadSubprocessPipeProto fd=2 pipe=<_UnixReadPipeTransport fd=14 polling>>)
INFO:asyncio:execute program 'python': <_UnixSubprocessTransport pid=92428 running stdin=<_UnixWritePipeTransport fd=8 idle bufsize=0> stdout=<_UnixReadPipeTransport fd=9 polling> stderr=<_UnixReadPipeTransport fd=11 polling>>
INFO:asyncio:execute program 'python': <_UnixSubprocessTransport pid=92430 running stdin=<_UnixWritePipeTransport fd=10 idle bufsize=0> stdout=<_UnixReadPipeTransport fd=12 polling> stderr=<_UnixReadPipeTransport fd=14 polling>>
-2: player_1 cannot overlap factory placement. Existing factory at (42, 29) already.
100: player_0 lost all factories
INFO:asyncio:<_UnixReadPipeTransport fd=9 polling> was closed by peer
DEBUG:asyncio:process 92428 exited with returncode -15
INFO:asyncio:<_UnixWritePipeTransport fd=8 idle bufsize=0> was closed by peer
INFO:asyncio:<_UnixSubprocessTransport pid=92428 running stdin=<_UnixWritePipeTransport closed fd=8 closed> stdout=<_UnixReadPipeTransport closed fd=9 closed> stderr=<_UnixReadPipeTransport fd=11 polling>> exited with return code -15
INFO:asyncio:<_UnixReadPipeTransport fd=12 polling> was closed by peer
DEBUG:asyncio:process 92430 exited with returncode -15
INFO:asyncio:<_UnixWritePipeTransport fd=10 idle bufsize=0> was closed by peer
INFO:asyncio:<_UnixSubprocessTransport pid=92430 running stdin=<_UnixWritePipeTransport closed fd=10 closed> stdout=<_UnixReadPipeTransport closed fd=12 closed> stderr=<_UnixReadPipeTransport fd=14 polling>> exited with return code -15
DEBUG:asyncio:Close <_UnixSelectorEventLoop running=False closed=False debug=True>
1.5513412952423096
/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py:125: ResourceWarning: unclosed transport <_UnixSubprocessTransport pid=92428 returncode=-15 stdin=<_UnixWritePipeTransport closed fd=8 closed> stdout=<_UnixReadPipeTransport closed fd=9 closed> stderr=<_UnixReadPipeTransport fd=11 open>>
  _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
Object allocated at (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", lineno 197
    transp = _UnixSubprocessTransport(self, protocol, args, shell,
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x7f0282baa5e0>
Traceback (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 126, in __del__
    self.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 104, in close
    proto.pipe.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 536, in close
    self._close(None)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 560, in _close
    self._loop.call_soon(self._call_connection_lost, exc)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py:125: ResourceWarning: unclosed transport <_UnixSubprocessTransport pid=92430 returncode=-15 stdin=<_UnixWritePipeTransport closed fd=10 closed> stdout=<_UnixReadPipeTransport closed fd=12 closed> stderr=<_UnixReadPipeTransport fd=14 open>>
  _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
Object allocated at (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", lineno 197
    transp = _UnixSubprocessTransport(self, protocol, args, shell,
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x7f0282baa5e0>
Traceback (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 126, in __del__
    self.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_subprocess.py", line 104, in close
    proto.pipe.close()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 536, in close
    self._close(None)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", line 560, in _close
    self._loop.call_soon(self._call_connection_lost, exc)
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py:540: ResourceWarning: unclosed transport <_UnixReadPipeTransport closing fd=11 open>
  _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
Object allocated at (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", lineno 179
    return _UnixReadPipeTransport(self, pipe, protocol, waiter, extra)
/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py:540: ResourceWarning: unclosed transport <_UnixReadPipeTransport closing fd=14 open>
  _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
Object allocated at (most recent call last):
  File "/home/jasper/.miniconda3/envs/luxai2022/lib/python3.8/asyncio/unix_events.py", lineno 179
    return _UnixReadPipeTransport(self, pipe, protocol, waiter, extra)
```

Changing the `cleanup` method to the following fixes it on my system:
```python
    async def cleanup(self):
        self._agent_process._transport.close()
        self._agent_process.terminate()
        await self._agent_process.wait()
```

Can you check whether this change also works on Windows and doesn't break the fix for #80?

Coincidentally this PR also fixes an issue in the tournament runner for me. Without the changes in this PR the tournament runner doesn't seem to terminate old agent processes, so it keeps spawning new agent processes until you run out of memory.

Comment (AmonUCF) [reply luxai_runner/process.py:100]: Added the recommended change and it's still working to get rid of the verbose error logs.

**Review by jmerle — APPROVED** [review]

**Comment by StoneT2000** [comment]
Thank you @AmonUCF @jmerle we'll take a look on our side hopefully soon, this will go into the 1.1.0 release #100

**Review by StoneT2000 — APPROVED** [review]

**Pull request merged**